### ISSUE TITLE
Run Once containers vs. Chaining / Container dependencies

### ISSUE THREAD

**Issue by markturansky** [GoogleCloudPlatform/kubernetes#1996]
Some commentary on "run once" containers:  #1589 #1515 

#### tl;dr

"run once" containers imply a pod lifecycle.  Declaring container dependencies ignores pod lifecycle and allows arbitrary complexity and flexibility for application developers ("run many").

#### Run Once

Add an attribute to the pod manifest containing either one or a list of docker images to run before the pod's main containers.  This is a single task list to run during pod creation, making this a lifecycle step. Implied, I think, is success of all tasks for a pod to continue. Also implied is linear processing because some tasks will require order for setup/config/init/whatever while others won't, but we cannot know which without understanding their dependencies.  Only some tasks might be required to complete before some of the pod's containers can be created.  I don't think this is a concurrent design.

#### Chaining / Container dependencies

A "depends" attribute on the container listing the docker images to run before the current container.  These are multiple task lists/chains before container creation allowing for greater flexibility for application developers/composers.  Containers can be created and come online as their dependencies are satisfied.  I think this way is the concurrent design and allows arbitrary complexity of composition for the application developer (including those tasks that must be sequential).

```
{
    "apiVersion": "v1beta4",
    "kind": "Pod",
    "id": "myapp",
    "desiredState": {
        "manifest": {
            "version": "v1beta4",
            "id": "myapp",
            "containers": [
                {
                    "name": "some-init",
                    "image": "myapp/something-init"
                },
                {
                    "name": "a-task",
                    "image": "myapp/random-task"
                },
                {
                    "name": "main-app",
                    "image": "myapp/main-app",
                    "depends": ["some-init", "a-task"]
                },
                {
                    "name": "another-container",
                    "image": "myapp/main-app",
                    "depends": ["a-task"]
                }
            ]
        }
    }
}
```

**Comment by bgrant0607** [comment]
General inter-container dependencies don't have well-defined semantics on updates and restarts, and would require checkpoints in Kubelet to implement correctly across Kubelet restarts, which we've so far avoided. Additionally, in my experience, this is a very slippery slope with a potentially very wide feature set for things like conditional triggers and error handling. I don't want BPEL in the PodSpec. 

I'd prefer to support the most common cases with more constrained mechanisms, and provide adequate extension points for users to implement more sophisticated behavior if they need to.

**Comment by markturansky** [comment]
no need for this issue, then.  I'll close it.
